cirq/protocols/inverse.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any
+
+import collections
+
+
+# This is a special indicator value used by the inverse method to determine
+# whether or not the caller provided a 'default' argument.
+RaiseTypeErrorIfNotProvided = ([],)
+
+
+def inverse(val: Any, default: Any = RaiseTypeErrorIfNotProvided) -> Any:
+    """Returns the inverse `val**-1` of the given value, if defined.
+
+    An object can define an inverse by defining a __pow__(self, exponent) method
+    that returns something besides NotImplemented when given the exponent -1.
+    The inverse of iterables is by default defined to be the iterable's items,
+    each inverted, in reverse order.
+
+    Args:
+        val: The value (or iterable of invertable values) to invert.
+        default: Determines the fallback behavior when `val` doesn't have
+            an inverse defined. If `default` is not set, a TypeError is raised.
+            If `default` is set to a value, that value is returned.
+
+    Returns:
+        If `val` has a __pow__ method that returns something besides
+        NotImplemented when given an exponent of -1, that result is returned.
+        Otherwise, if `val` is iterable, the result is a tuple with the same
+        items as `val` but in reverse order and with each item inverted.
```

Comment:
we have cases where the iterable case is used?

Reply:
`cirq.inverse` used to be something like `inverse_of_op_tree`, so actually all existing uses were on the iterable case.